Mathlib/RingTheory/Polynomial/Hilbert.lean

```diff
@@ -0,0 +1,199 @@
+  · left
+    rw [← Nat.cast_div (Nat.factorial_dvd_factorial <| Nat.le_add_right (n - k) d) <| by
+      simp only [ne_eq, Nat.cast_eq_zero]; exact Nat.factorial_ne_zero (n - k),
+      ← Nat.ascFactorial_eq_div]
+    induction d with
+    | zero => simp only [Nat.zero_eq, Finset.range_zero, Finset.prod_empty, Nat.ascFactorial_zero,
+        Nat.cast_one]
+    | succ d hd => rw [Finset.prod_range_succ, hd, add_comm d, mul_comm, eval_natCast,
+        ← Nat.cast_sub hkn, eval_natCast, ← Nat.cast_add, ← Nat.cast_add_one, ← Nat.cast_mul,
+        ← Nat.succ_eq_one_add, Nat.ascFactorial_succ, add_assoc, add_comm d, ← add_assoc]
+
+/--
+Given `p : ℤ[X]` and `d : ℕ`, the Hilbert polynomial of `p` and `d`.
+Look at `Polynomial.coeff_mul_invOneSubPow_eq_hilbert_eval`, which says
+that `PowerSeries.coeff ℤ n (p * (@invOneSubPow ℤ _ d))` is equal to
+`(Polynomial.hilbert p d).eval (n : ℚ)` for any large enough `n : ℕ`.
+-/
+noncomputable def hilbert (p : ℤ[X]) (d : ℕ) : ℚ[X] :=
+  if h : p = 0 then 0
+  else if d + 1 ≤ p.rootMultiplicity 1 then 0
+  else ∑ i in Finset.range ((greatestFactorOneSubNotDvd p h).natDegree + 1),
+  ((greatestFactorOneSubNotDvd p h).coeff i) • preHilbert (d - p.rootMultiplicity 1) i
+
+/--
+Given `p : ℤ[X]` and `d : ℕ`. The key property of the Hilbert polynomial with respect to
+`p` and `d`, which says that for any term of `p * (@invOneSubPow ℤ _ d)` whose degree is
+large enough, its coefficient can be obtained by evaluating the Hilbert polynomial.
+-/
+theorem coeff_mul_invOneSubPow_eq_hilbert_eval (p : ℤ[X]) (d n : ℕ) (hn : p.natDegree < n) :
+    PowerSeries.coeff ℤ n (p * (@invOneSubPow ℤ _ d)) = (hilbert p d).eval (n : ℚ) := by
+  rw [hilbert]; by_cases h : p = 0
+  · simp only [h, coe_zero, zero_mul, ↓reduceDIte, eval_zero, Int.cast_eq_zero]; rfl
+  · simp only [h, ↓reduceDIte, zsmul_eq_mul]
+    have coe_one_sub : (1 - X : ℤ[X]).ToPowerSeries = 1 - (PowerSeries.X : ℤ⟦X⟧) :=
+      PowerSeries.ext_iff.2 fun i => by_cases (fun (hi : i = 0) => by
+      simp only [hi, coeff_coe, coeff_sub, coeff_one_zero, coeff_X_zero, sub_zero,
+      map_sub, PowerSeries.coeff_one, ↓reduceIte, coeff_zero_X]) (fun hi => by
+      simp only [coeff_coe, coeff_sub, map_sub, PowerSeries.coeff_one, hi, ↓reduceIte,
+      zero_sub]; rw [Polynomial.coeff_one]; simp only [hi, ↓reduceIte, zero_sub, neg_inj];
+      rw [Polynomial.coeff_X, PowerSeries.coeff_X]; exact by_cases (fun (hi : i = 1) => by
+      simp only [hi]) (fun hi => by simp only [hi]; exact if_neg <| Ne.symm hi))
+    by_cases h1 : d + 1 ≤ p.rootMultiplicity 1
+    · simp only [h1, ↓reduceIte, eval_zero, Int.cast_eq_zero]
+      rw [← pow_rootMultiplicity_mul_greatestFactorOneSubNotDvd_eq p h, mul_comm, coe_mul,
+        ← mul_assoc, coe_pow, coe_one_sub, ← @Nat.sub_add_cancel (p.rootMultiplicity 1)
+        (d + 1) h1, mul_comm (invOneSubPow d).val, pow_add, mul_assoc ((1 - PowerSeries.X) ^
+        (p.rootMultiplicity 1 - (d + 1)))]
+      erw [(invOneSubPow d).inv_val]
```

Comment:
This erw should be removed (because right now the culture is to argue that they're evil). You should add the lemma saying that (1-X)^(d+1)*invOneSubPow d = 1 in the invOneSubPow file and rewrite with this instead.